// react-next-tilt/dist/react-next-tilt.js
import mr, { forwardRef as pr, useRef as Ye, useCallback as _, useMemo as fr, useImperativeHandle as hr, useEffect as sr, memo as yr } from "react";
var or = { exports: {} }, qe = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var cr;
function br() {
  if (cr)
    return qe;
  cr = 1;
  var a = mr, f = Symbol.for("react.element"), y = Symbol.for("react.fragment"), v = Object.prototype.hasOwnProperty, b = a.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, E = { key: !0, ref: !0, __self: !0, __source: !0 };
  function re(F, m, ue) {
    var T, U = {}, j = null, we = null;
    ue !== void 0 && (j = "" + ue), m.key !== void 0 && (j = "" + m.key), m.ref !== void 0 && (we = m.ref);
    for (T in m)
      v.call(m, T) && !E.hasOwnProperty(T) && (U[T] = m[T]);
    if (F && F.defaultProps)
      for (T in m = F.defaultProps, m)
        U[T] === void 0 && (U[T] = m[T]);
    return { $$typeof: f, type: F, key: j, ref: we, props: U, _owner: b.current };
  }
  return qe.Fragment = y, qe.jsx = re, qe.jsxs = re, qe;
}
var Be = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var ir;
function gr() {
  return ir || (ir = 1, process.env.NODE_ENV !== "production" && function() {
    var a = mr, f = Symbol.for("react.element"), y = Symbol.for("react.portal"), v = Symbol.for("react.fragment"), b = Symbol.for("react.strict_mode"), E = Symbol.for("react.profiler"), re = Symbol.for("react.provider"), F = Symbol.for("react.context"), m = Symbol.for("react.forward_ref"), ue = Symbol.for("react.suspense"), T = Symbol.for("react.suspense_list"), U = Symbol.for("react.memo"), j = Symbol.for("react.lazy"), we = Symbol.for("react.offscreen"), Ie = Symbol.iterator, ye = "@@iterator";
    function Pe(e) {
      if (e === null || typeof e != "object")
        return null;
      var r = Ie && e[Ie] || e[ye];
      return typeof r == "function" ? r : null;
    }
    var te = a.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function p(e) {
      {
        for (var r = arguments.length, t = new Array(r > 1 ? r - 1 : 0), u = 1; u < r; u++)
          t[u - 1] = arguments[u];
        He("error", e, t);
      }
    }
    function He(e, r, t) {
      {
        var u = te.ReactDebugCurrentFrame, l = u.getStackAddendum();
        l !== "" && (r += "%s", t = t.concat([l]));
        var d = t.map(function(c) {
          return String(c);
        });
        d.unshift("Warning: " + r), Function.prototype.apply.call(console[e], console, d);
      }
    }
    var Te = !1, Le = !1, Me = !1, be = !1, Ge = !1, We;
    We = Symbol.for("react.module.reference");
    function Z(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === v || e === E || Ge || e === b || e === ue || e === T || be || e === we || Te || Le || Me || typeof e == "object" && e !== null && (e.$$typeof === j || e.$$typeof === U || e.$$typeof === re || e.$$typeof === F || e.$$typeof === m || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === We || e.getModuleId !== void 0));
    }
    function Q(e, r, t) {
      var u = e.displayName;
      if (u)
        return u;
      var l = r.displayName || r.name || "";
      return l !== "" ? t + "(" + l + ")" : t;
    }
    function q(e) {
      return e.displayName || "Context";
    }
    function Y(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && p("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case v:
          return "Fragment";
        case y:
          return "Portal";
        case E:
          return "Profiler";
        case b:
          return "StrictMode";
        case ue:
          return "Suspense";
        case T:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case F:
            var r = e;
            return q(r) + ".Consumer";
          case re:
            var t = e;
            return q(t._context) + ".Provider";
          case m:
            return Q(e, e.render, "ForwardRef");
          case U:
            var u = e.displayName || null;
            return u !== null ? u : Y(e.type) || "Memo";
          case j: {
            var l = e, d = l._payload, c = l._init;
            try {
              return Y(c(d));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var I = Object.assign, C = 0, L, B, Je, ze, Ke, Ae, H;
    function J() {
    }
    J.__reactDisabledLog = !0;
    function Ze() {
      {
        if (C === 0) {
          L = console.log, B = console.info, Je = console.warn, ze = console.error, Ke = console.group, Ae = console.groupCollapsed, H = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: J,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        C++;
      }
    }
    function w() {
      {
        if (C--, C === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: I({}, e, {
              value: L
            }),
            info: I({}, e, {
              value: B
            }),
            warn: I({}, e, {
              value: Je
            }),
            error: I({}, e, {
              value: ze
            }),
            group: I({}, e, {
              value: Ke
            }),
            groupCollapsed: I({}, e, {
              value: Ae
            }),
            groupEnd: I({}, e, {
              value: H
            })
          });
        }
        C < 0 && p("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var Ne = te.ReactCurrentDispatcher, fe;
    function Ce(e, r, t) {
      {
        if (fe === void 0)
          try {
            throw Error();
          } catch (l) {
            var u = l.stack.trim().match(/\n( *(at )?)/);
            fe = u && u[1] || "";
          }
        return `
` + fe + e;
      }
    }
    var k = !1, S;
    {
      var er = typeof WeakMap == "function" ? WeakMap : Map;
      S = new er();
    }
    function Ve(e, r) {
      if (!e || k)
        return "";
      {
        var t = S.get(e);
        if (t !== void 0)
          return t;
      }
      var u;
      k = !0;
      var l = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var d;
      d = Ne.current, Ne.current = null, Ze();
      try {
        if (r) {
          var c = function() {
            throw Error();
          };
          if (Object.defineProperty(c.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(c, []);
            } catch (ae) {
              u = ae;
            }
            Reflect.construct(e, [], c);
          } else {
            try {
              c.call();
            } catch (ae) {
              u = ae;
            }
            e.call(c.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (ae) {
            u = ae;
          }
          e();
        }
      } catch (ae) {
        if (ae && u && typeof ae.stack == "string") {
          for (var s = ae.stack.split(`
`), O = u.stack.split(`
`), h = s.length - 1, g = O.length - 1; h >= 1 && g >= 0 && s[h] !== O[g]; )
            g--;
          for (; h >= 1 && g >= 0; h--, g--)
            if (s[h] !== O[g]) {
              if (h !== 1 || g !== 1)
                do
                  if (h--, g--, g < 0 || s[h] !== O[g]) {
                    var X = `
` + s[h].replace(" at new ", " at ");
                    return e.displayName && X.includes("<anonymous>") && (X = X.replace("<anonymous>", e.displayName)), typeof e == "function" && S.set(e, X), X;
                  }
                while (h >= 1 && g >= 0);
              break;
            }
        }
      } finally {
        k = !1, Ne.current = d, w(), Error.prepareStackTrace = l;
      }
      var De = e ? e.displayName || e.name : "", ur = De ? Ce(De) : "";
      return typeof e == "function" && S.set(e, ur), ur;
    }
    function se(e, r, t) {
      return Ve(e, !1);
    }
    function xe(e) {
      var r = e.prototype;
      return !!(r && r.isReactComponent);
    }
    function ce(e, r, t) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return Ve(e, xe(e));
      if (typeof e == "string")
        return Ce(e);
      switch (e) {
        case ue:
          return Ce("Suspense");
        case T:
          return Ce("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case m:
            return se(e.render);
          case U:
            return ce(e.type, r, t);
          case j: {
            var u = e, l = u._payload, d = u._init;
            try {
              return ce(d(l), r, t);
            } catch {
            }
          }
        }
      return "";
    }
    var ie = Object.prototype.hasOwnProperty, Oe = {}, je = te.ReactDebugCurrentFrame;
    function le(e) {
      if (e) {
        var r = e._owner, t = ce(e.type, e._source, r ? r.type : null);
        je.setExtraStackFrame(t);
      } else
        je.setExtraStackFrame(null);
    }
    function Xe(e, r, t, u, l) {
      {
        var d = Function.call.bind(ie);
        for (var c in e)
          if (d(e, c)) {
            var s = void 0;
            try {
              if (typeof e[c] != "function") {
                var O = Error((u || "React class") + ": " + t + " type `" + c + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[c] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw O.name = "Invariant Violation", O;
              }
              s = e[c](r, c, u, t, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (h) {
              s = h;
            }
            s && !(s instanceof Error) && (le(l), p("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", u || "React class", t, c, typeof s), le(null)), s instanceof Error && !(s.message in Oe) && (Oe[s.message] = !0, le(l), p("Failed %s type: %s", t, s.message), le(null));
          }
      }
    }
    var Fe = Array.isArray;
    function Ue(e) {
      return Fe(e);
    }
    function rr(e) {
      {
        var r = typeof Symbol == "function" && Symbol.toStringTag, t = r && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return t;
      }
    }
    function tr(e) {
      try {
        return R(e), !1;
      } catch {
        return !0;
      }
    }
    function R(e) {
      return "" + e;
    }
    function ve(e) {
      if (tr(e))
        return p("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", rr(e)), R(e);
    }
    var W = te.ReactCurrentOwner, P = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, N, $, V;
    V = {};
    function nr(e) {
      if (ie.call(e, "ref")) {
        var r = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function ar(e) {
      if (ie.call(e, "key")) {
        var r = Object.getOwnPropertyDescriptor(e, "key").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function ke(e, r) {
      if (typeof e.ref == "string" && W.current && r && W.current.stateNode !== r) {
        var t = Y(W.current.type);
        V[t] || (p('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', Y(W.current.type), e.ref), V[t] = !0);
      }
    }
    function Qe(e, r) {
      {
        var t = function() {
          N || (N = !0, p("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        t.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: t,
          configurable: !0
        });
      }
    }
    function D(e, r) {
      {
        var t = function() {
          $ || ($ = !0, p("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        t.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: t,
          configurable: !0
        });
      }
    }
    var x = function(e, r, t, u, l, d, c) {
      var s = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: f,
        // Built-in properties that belong on the element
        type: e,
        key: r,
        ref: t,
        props: c,
        // Record the component responsible for creating this element.
        _owner: d
      };
      return s._store = {}, Object.defineProperty(s._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(s, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: u
      }), Object.defineProperty(s, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: l
      }), Object.freeze && (Object.freeze(s.props), Object.freeze(s)), s;
    };
    function de(e, r, t, u, l) {
      {
        var d, c = {}, s = null, O = null;
        t !== void 0 && (ve(t), s = "" + t), ar(r) && (ve(r.key), s = "" + r.key), nr(r) && (O = r.ref, ke(r, l));
        for (d in r)
          ie.call(r, d) && !P.hasOwnProperty(d) && (c[d] = r[d]);
        if (e && e.defaultProps) {
          var h = e.defaultProps;
          for (d in h)
            c[d] === void 0 && (c[d] = h[d]);
        }
        if (s || O) {
          var g = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          s && Qe(c, g), O && D(c, g);
        }
        return x(e, s, O, l, u, W.current, c);
      }
    }
    var M = te.ReactCurrentOwner, G = te.ReactDebugCurrentFrame;
    function ne(e) {
      if (e) {
        var r = e._owner, t = ce(e.type, e._source, r ? r.type : null);
        G.setExtraStackFrame(t);
      } else
        G.setExtraStackFrame(null);
    }
    var ee;
    ee = !1;
    function z(e) {
      return typeof e == "object" && e !== null && e.$$typeof === f;
    }
    function K() {
      {
        if (M.current) {
          var e = Y(M.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function Se(e) {
      {
        if (e !== void 0) {
          var r = e.fileName.replace(/^.*[\\\/]/, ""), t = e.lineNumber;
          return `

Check your code at ` + r + ":" + t + ".";
        }
        return "";
      }
    }
    var me = {};
    function ge(e) {
      {
        var r = K();
        if (!r) {
          var t = typeof e == "string" ? e : e.displayName || e.name;
          t && (r = `

Check the top-level render call using <` + t + ">.");
        }
        return r;
      }
    }
    function pe(e, r) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var t = ge(r);
        if (me[t])
          return;
        me[t] = !0;
        var u = "";
        e && e._owner && e._owner !== M.current && (u = " It was passed a child from " + Y(e._owner.type) + "."), ne(e), p('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', t, u), ne(null);
      }
    }
    function he(e, r) {
      {
        if (typeof e != "object")
          return;
        if (Ue(e))
          for (var t = 0; t < e.length; t++) {
            var u = e[t];
            z(u) && pe(u, r);
          }
        else if (z(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var l = Pe(e);
          if (typeof l == "function" && l !== e.entries)
            for (var d = l.call(e), c; !(c = d.next()).done; )
              z(c.value) && pe(c.value, r);
        }
      }
    }
    function Ee(e) {
      {
        var r = e.type;
        if (r == null || typeof r == "string")
          return;
        var t;
        if (typeof r == "function")
          t = r.propTypes;
        else if (typeof r == "object" && (r.$$typeof === m || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        r.$$typeof === U))
          t = r.propTypes;
        else
          return;
        if (t) {
          var u = Y(r);
          Xe(t, e.props, "prop", u, e);
        } else if (r.PropTypes !== void 0 && !ee) {
          ee = !0;
          var l = Y(r);
          p("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", l || "Unknown");
        }
        typeof r.getDefaultProps == "function" && !r.getDefaultProps.isReactClassApproved && p("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function _e(e) {
      {
        for (var r = Object.keys(e.props), t = 0; t < r.length; t++) {
          var u = r[t];
          if (u !== "children" && u !== "key") {
            ne(e), p("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", u), ne(null);
            break;
          }
        }
        e.ref !== null && (ne(e), p("Invalid attribute `ref` supplied to `React.Fragment`."), ne(null));
      }
    }
    function $e(e, r, t, u, l, d) {
      {
        var c = Z(e);
        if (!c) {
          var s = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (s += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var O = Se(l);
          O ? s += O : s += K();
          var h;
          e === null ? h = "null" : Ue(e) ? h = "array" : e !== void 0 && e.$$typeof === f ? (h = "<" + (Y(e.type) || "Unknown") + " />", s = " Did you accidentally export a JSX literal instead of a component?") : h = typeof e, p("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", h, s);
        }
        var g = de(e, r, t, l, d);
        if (g == null)
          return g;
        if (c) {
          var X = r.children;
          if (X !== void 0)
            if (u)
              if (Ue(X)) {
                for (var De = 0; De < X.length; De++)
                  he(X[De], e);
                Object.freeze && Object.freeze(X);
              } else
                p("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              he(X, e);
        }
        return e === v ? _e(g) : Ee(g), g;
      }
    }
    function n(e, r, t) {
      return $e(e, r, t, !0);
    }
    function o(e, r, t) {
      return $e(e, r, t, !1);
    }
    var i = o, Re = n;
    Be.Fragment = v, Be.jsx = i, Be.jsxs = Re;
  }()), Be;
}
process.env.NODE_ENV === "production" ? or.exports = br() : or.exports = gr();
var oe = or.exports;
const A = (a, f, y) => a < f ? f : a > y ? y : a, Er = (a, f, y) => {
  let v = 0;
  return f === "top" ? v = (a.offsetY * 2 - 1) * -1 : f === "bottom" ? v = (a.offsetY - 0.5) * 2 : f === "left" ? v = (a.offsetX * 2 - 1) * -1 : f === "right" ? v = (a.offsetX - 0.5) * 2 : f === "all" && (v = Math.max(
    Math.abs(a.offsetX - 0.5) * 2,
    Math.abs(a.offsetY - 0.5) * 2
  )), (v * y).toFixed(2);
}, _r = (a, f, y) => {
  let v = a.offsetX, b = a.offsetY;
  y || (v = 1 - v, b = 1 - b);
  let E = "translateX(0%) translateY(0%)";
  return f === "top" ? E = `translateX( ${v * 50}% )` : f === "bottom" ? E = `translateX( ${v * 50}% ) translateY(50%)` : f === "left" ? E = `translateY( ${b * 50}% )` : f === "right" ? E = `translateX(50%) translateY( ${b * 50}% ) ` : f === "all" && (E = `translateX(${(v - 0.5) * 50}%) translateY(${(b - 0.5) * 50}%) `), E;
}, lr = (a) => {
  if (!a)
    return "translateX(0%) translateY(0%)";
  switch (a) {
    case "top-left":
      a = { left: "25%", top: "25%" };
      break;
    case "top-right":
      a = { left: "75%", top: "25%" };
      break;
    case "bottom-left":
      a = { left: "25%", top: "75%" };
      break;
    case "bottom-right":
      a = { left: "75%", top: "75%" };
      break;
    case "center":
      a = { left: "50%", top: "50%" };
      break;
  }
  const f = a.left.includes("%") ? String(parseInt(a.left) / 2) + "%" : a.left, y = a.top.includes("%") ? String(parseInt(a.top) / 2) + "%" : a.top;
  return `translateX(${f}) translateY(${y})`;
}, Rr = (a, f, y) => {
  let v = a.offsetX, b = a.offsetY;
  f === "top-right" && (v = 1 - v), f === "bottom-left" && (b = 1 - b), f === "bottom-right" && (v = 1 - v, b = 1 - b);
  let E = 0;
  return y ? E = ((v + b) * (3 / 2) - 1) * 100 : E = ((v + b) * (-3 / 2) + 0.5) * 100, `translateX(${E}%)`;
}, vr = (a, f) => {
  if (!f)
    return `translateX(${a ? "50%" : "-100%"})`;
  switch (f) {
    case "left":
      f = { left: "25%" };
      break;
    case "center":
      f = { left: "50%" };
      break;
    case "right":
      f = { left: "75%" };
      break;
  }
  return `translateX(${f.left.includes("%") ? String(parseInt(f.left) / 2) + "%" : f.left})`;
}, dr = (a) => {
  if (a instanceof HTMLElement)
    return a;
  if (a instanceof Document)
    return document.documentElement;
  if (a.current && a.current instanceof Object && "element" in a.current && a.current.element instanceof HTMLElement)
    return a.current.element;
  if (a.current && a.current instanceof HTMLElement)
    return a.current;
}, wr = pr(
  ({
    width: a,
    height: f,
    borderRadius: y,
    perspective: v = "1000px",
    scale: b = 1,
    shadowEnable: E = !1,
    shadow: re = "0 0 1rem rgba(0,0,0,0.5)",
    shadowType: F = "box",
    lineGlareEnable: m = !0,
    lineGlareBlurEnable: ue = !0,
    lineGlareBlurAmount: T = "4px",
    lineGlareMaxOpacity: U = 0.1,
    lineGlareWidthPercent: j = 10,
    lineGlareMixBlendMode: we = "normal",
    lineGlareColor: Ie = "white",
    lineGlareReverse: ye = !1,
    lineGlareDirection: Pe = "to-bottom-right",
    lineGlareHoverPosition: te = "top-left",
    lineGlareFixedPosition: p = void 0,
    spotGlareEnable: He = !0,
    spotGlareSizePercent: Te = 200,
    spotGlareMaxOpacity: Le = 0.5,
    spotGlareMixBlendMode: Me = "normal",
    spotGlarePosition: be = "top",
    spotGlareColor: Ge = "white",
    spotGlareReverse: We = !1,
    spotGlareFixedPosition: Z = void 0,
    tiltMaxAngleX: Q = 20,
    tiltMaxAngleY: q = 20,
    tiltReverse: Y = !1,
    tiltReset: I = !0,
    initialAngleX: C = 0,
    initialAngleY: L = 0,
    disableScrollOnTouch: B = !0,
    style: Je,
    tiltStyle: ze,
    tiltClass: Ke,
    tiltProps: Ae,
    gyroMaxAngleX: H = 0,
    gyroMaxAngleY: J = 0,
    gyroReverse: Ze = !1,
    disabled: w = !1,
    disabledFilter: Ne = "grayscale(1) brightness(125%)",
    CSSTransition: fe = "all 0.4s cubic-bezier(0.03, 0.98, 0.52, 0.99)",
    TiltWrapper: Ce,
    fullPageListening: k = !1,
    controlElement: S,
    controlElementOnly: er = !1,
    preserve3dEnable: Ve = !0,
    testIdEnable: se = !1,
    onTilt: xe,
    onReset: ce,
    onMouseEnter: ie,
    onMouseMove: Oe,
    onMouseLeave: je,
    onTouchStart: le,
    onTouchMove: Xe,
    onTouchEnd: Fe,
    children: Ue,
    ...rr
  }, tr) => {
    const R = Ye(), ve = Ye(!1), W = !k && (!S || S && !er), P = Ye(null), N = Ye(null), $ = Ye(null), V = Ye(null), nr = _(
      ({ children: n }) => /* @__PURE__ */ oe.jsx(oe.Fragment, { children: n }),
      []
    ), ar = Ce || nr;
    m && (j = A(j, 0, 50) / 2), He && (Te = Te < 0 ? 0 : Te / 2), Q = A(Q, 0, 90), q = A(q, 0, 90), H = A(H, 0, 90), J = A(J, 0, 90), C && (C = A(
      C,
      -Q,
      Q
    )), L && (L = A(
      L,
      -q,
      q
    ));
    const ke = fr(() => m ? (100 - j) / 2 : 0, [m, j]), Qe = fr(() => m ? ke + j : 0, [m, j, ke]), D = _((n = !0) => {
      requestAnimationFrame(() => {
        N.current && (N.current.style.willChange = n ? "transform" : ""), $.current && ($.current.style.willChange = n ? "transform, opacity" : ""), V.current && (V.current.style.willChange = n ? "transform" : "");
      });
    }, []), x = _(
      (n = !0) => {
        requestAnimationFrame(() => {
          if (F === "box")
            N.current && E && (N.current.style.boxShadow = n ? re : "");
          else if (P.current && E) {
            const o = P.current.style.filter.replace(
              /(drop-shadow?\(.*?\))(?=\s[a-z].*?\(.*?\)|\s*$)/g,
              ""
            );
            P.current.style.filter = n ? o + ` drop-shadow(${re})` : o;
          }
        });
      },
      [re, E, F]
    ), de = _(() => {
      if (!P.current || !$.current || !R.current || Z)
        return;
      const n = Er(
        R.current,
        be,
        Le
      ), o = _r(
        R.current,
        be,
        We
      );
      requestAnimationFrame(() => {
        $.current && ($.current.style.opacity = n, $.current.style.transform = o);
      });
    }, [
      be,
      Le,
      We,
      Z
    ]), M = _(() => {
      if (!P.current || !V.current || !R.current || p)
        return;
      const n = Rr(
        R.current,
        te,
        ye
      );
      requestAnimationFrame(() => {
        V.current && (V.current.style.transform = n);
      });
    }, [te, ye, p]), G = _(
      (n = !0) => {
        if (!P.current || !R.current)
          return { angleX: 0, angleY: 0 };
        let o = -(R.current.offsetY - 0.5) * 2 * Q, i = (R.current.offsetX - 0.5) * 2 * q;
        return Y && n && (o = -o, i = -i), { angleX: o, angleY: i };
      },
      [Q, q, Y]
    ), ne = _(
      (n) => {
        const o = (n.angleY / q + 1) / 2, i = (-n.angleX / Q + 1) / 2;
        R.current = { offsetX: o, offsetY: i };
      },
      [Q, q]
    ), ee = _(
      (n, o = !0, i = !1) => {
        const Re = o ? b : 1;
        requestAnimationFrame(() => {
          N.current && (N.current.style.transform = `rotateX(${n.angleX}deg) rotateY(${n.angleY}deg) scale3d(${Re},${Re},${Re})`);
        }), xe && xe(
          {
            angleX: n.angleX,
            angleY: n.angleY
          },
          i
        );
      },
      [xe, b]
    ), z = _(
      (n, o = !1, i = !1) => {
        ne(n), ee(n, o, i), x(o), M(), de();
      },
      [
        ne,
        ee,
        x,
        M,
        de
      ]
    ), K = _(() => {
      ce && ce(), C || L ? z({
        angleX: C || 0,
        angleY: L || 0
      }) : (R.current = void 0, requestAnimationFrame(() => {
        V.current && !p && (V.current.style.transform = vr(ye)), $.current && !Z && ($.current.style.transform = lr(), $.current.style.opacity = "0"), N.current && (N.current.style.transform = "rotateX(0deg) rotateY(0deg) scale3d(1, 1, 1)");
      }));
    }, [
      C,
      L,
      ye,
      Z,
      p,
      ce,
      z
    ]), Se = _(
      (n, o) => {
        if (!o)
          return;
        if (k) {
          R.current = {
            offsetX: n.x / window.innerWidth,
            offsetY: n.y / window.innerHeight
          };
          return;
        }
        const i = o.getBoundingClientRect();
        if (!i.width || !i.height)
          return;
        const Re = A(
          n.x,
          i.left,
          i.right
        ), e = A(
          n.y,
          i.top,
          i.bottom
        ), r = (Re - i.left) / i.width, t = (e - i.top) / i.height;
        R.current = { offsetX: r, offsetY: t };
      },
      [k]
    );
    hr(
      tr,
      () => ({
        tilt: z,
        reset: K,
        angle: () => G(!1),
        updateWillChange: D,
        element: P.current
      }),
      [z, K, D, G]
    );
    const me = _(() => {
      w || (ve.current = !0, D(), x());
    }, [w, x, D]), ge = _(() => {
      w || (B && !k && (typeof B == "boolean" ? document.body.style.overflow = "hidden" : B.style.overflow = "hidden"), ve.current = !0, D(), x());
    }, [
      B,
      w,
      k,
      x,
      D
    ]), pe = _(
      (n) => {
        w || (Se(
          { x: n.clientX, y: n.clientY },
          n.currentTarget
        ), ee(G()), M(), de());
      },
      [
        w,
        G,
        Se,
        ee,
        M,
        de
      ]
    ), he = _(
      (n) => {
        if (w)
          return;
        const o = n.changedTouches[0];
        o && (Se(
          {
            x: o.clientX,
            y: o.clientY
          },
          n.currentTarget
        ), ee(G()), M(), de());
      },
      [
        w,
        G,
        Se,
        ee,
        M,
        de
      ]
    ), Ee = _(() => {
      w || (ve.current = !1, D(!1), x(!1), I && K());
    }, [w, K, I, x, D]), _e = _(() => {
      w || (B && !k && (typeof B == "boolean" ? document.body.style.overflow = "" : B.style.overflow = ""), ve.current = !1, D(!1), x(!1), I && K());
    }, [
      B,
      w,
      k,
      K,
      I,
      x,
      D
    ]);
    sr(() => {
      if (!S && !k)
        return;
      let n;
      k || !S ? n = [document] : n = Array.isArray(S) ? S : [S];
      for (let o of n) {
        const i = dr(o);
        i && (o = i, o.addEventListener("mouseenter", me), o.addEventListener("mousemove", pe), o.addEventListener("mouseleave", Ee), o.addEventListener("touchstart", ge, { passive: !0 }), o.addEventListener("touchmove", he, { passive: !0 }), o.addEventListener("touchend", _e, { passive: !0 }));
      }
      return () => {
        for (let o of n) {
          const i = dr(o);
          i && (o = i, o.removeEventListener("mouseenter", me), o.removeEventListener("mousemove", pe), o.removeEventListener("mouseleave", Ee), o.removeEventListener("touchstart", ge), o.removeEventListener("touchmove", he), o.removeEventListener("touchend", _e));
        }
      };
    }, [
      S,
      me,
      pe,
      Ee,
      ge,
      he,
      _e,
      k
    ]);
    const $e = _(
      (n) => {
        if (ve.current || n.beta === null && n.gamma === null)
          return;
        let o = H && n.beta ? A(-n.beta, -H, H) : C || 0, i = J && n.gamma ? A(n.gamma, -J, J) : L || 0;
        Ze && (o = -o, i = -i), z({ angleX: o, angleY: i }, !1, !0);
      },
      [
        H,
        J,
        C,
        L,
        Ze,
        z
      ]
    );
    return sr(() => {
      if (!(!H && !J))
        return window.addEventListener("deviceorientation", $e), () => window.removeEventListener("deviceorientation", $e);
    }, [H, J, $e, D]), /* @__PURE__ */ oe.jsx(
      "div",
      {
        ref: (n) => {
          n && (P.current = n, R.current && (w ? K() : z(G())), (C || L) && !R.current && K());
        },
        "data-testid": se ? "container" : void 0,
        style: Object.assign(
          {
            display: "inline-block",
            verticalAlign: "middle",
            width: a,
            height: f,
            borderRadius: y,
            perspective: v,
            transformStyle: Ve ? "preserve-3d" : void 0,
            backfaceVisibility: "hidden",
            filter: w ? Ne : void 0,
            transition: F === "drop" ? fe : void 0
          },
          Je
        ),
        onMouseEnter: W ? ie ? (n) => {
          me(), ie(n);
        } : me : ie,
        onTouchStart: W ? le ? (n) => {
          ge(), le(n);
        } : ge : le,
        onMouseMove: W ? Oe ? (n) => {
          pe(n), Oe(n);
        } : pe : Oe,
        onTouchMove: W ? Xe ? (n) => {
          he(n), Xe(n);
        } : he : Xe,
        onMouseLeave: W ? je ? (n) => {
          Ee(), je(n);
        } : Ee : je,
        onTouchEnd: W ? Fe ? (n) => {
          _e(), Fe(n);
        } : _e : Fe,
        ...rr,
        children: /* @__PURE__ */ oe.jsx(ar, { children: /* @__PURE__ */ oe.jsxs(
          "div",
          {
            ref: N,
            "data-testid": se ? "tilt" : void 0,
            style: Object.assign(
              {
                boxSizing: "border-box",
                position: "relative",
                width: "100%",
                height: "100%",
                borderRadius: y,
                transformStyle: Ve ? "preserve-3d" : void 0,
                backfaceVisibility: "hidden",
                transition: fe,
                transform: "rotateX(0deg) rotateY(0deg) scale3d(1, 1, 1)"
              },
              ze
            ),
            className: Ke,
            ...Ae,
            children: [
              Ue,
              He && /* @__PURE__ */ oe.jsx(
                "div",
                {
                  "data-testid": se ? "spot-glare-container" : void 0,
                  style: {
                    pointerEvents: "none",
                    position: "absolute",
                    inset: "0",
                    borderRadius: y,
                    overflow: "hidden",
                    backfaceVisibility: "hidden",
                    mixBlendMode: Me
                  },
                  children: /* @__PURE__ */ oe.jsx(
                    "div",
                    {
                      ref: $,
                      "data-testid": se ? "spot-glare" : void 0,
                      style: {
                        position: "absolute",
                        left: be === "all" && !Z ? "-50%" : "-100%",
                        top: be === "all" && !Z ? "-50%" : "-100%",
                        width: "200%",
                        height: "200%",
                        transition: fe,
                        backgroundImage: `radial-gradient(${Ge}, transparent ${Te}%)`,
                        transform: lr(
                          Z
                        ),
                        opacity: Z ? Le : "0"
                      }
                    }
                  )
                }
              ),
              m && /* @__PURE__ */ oe.jsx(
                "div",
                {
                  "data-testid": se ? "line-glare-container" : void 0,
                  style: {
                    pointerEvents: "none",
                    position: "absolute",
                    inset: "0",
                    borderRadius: y,
                    overflow: "hidden",
                    backfaceVisibility: "hidden",
                    mixBlendMode: we
                  },
                  children: /* @__PURE__ */ oe.jsx(
                    "div",
                    {
                      "data-testid": se ? "line-glare" : void 0,
                      ref: V,
                      style: {
                        position: "absolute",
                        left: p ? "-100%" : 0,
                        top: "-50%",
                        width: "200%",
                        height: "200%",
                        filter: ue ? `blur(${T})` : "",
                        transition: fe,
                        opacity: String(U),
                        transform: vr(
                          ye,
                          p
                        ),
                        backgroundImage: `linear-gradient(${Pe === "to-bottom-right" ? "to bottom right" : "to bottom left"}, transparent, transparent ${ke}%, ${Ie} calc(${ke}% + 1px), ${Ie} ${Qe}%, transparent calc(${Qe}% + 1px), transparent 0%)`
                      }
                    }
                  )
                }
              )
            ]
          }
        ) })
      }
    );
  }
), Tr = yr(wr);
Tr.displayName = "Tilt";
export {
  Tr as Tilt,
  Tr as default
};
